azurelinuxagent/common/cgroups.py

```diff
@@ -538,9 +560,6 @@ def set_limits(self):
                 message=msg,
                 log_event=False)
 
```

Comment:
Just move the creation of self.threshold into the CGroup constructor and it will always be present, even if that particular instance of CGroup object was never used to set the thresholds.

Reply:
Do you mean:
1. assign the value in the constructor itself (compute the memory/CPU limits in the constructor and assign, which would be defaults)?
2. I could change this in the constructor of CGroup class - changing `__init__(self, name, path_maker)` to `__init__(self, name, path_maker, limits=None)`, and use it to be populated by the `for_extension(name, limits=None)` call, and if the limits is None, then use approach 1.
3. just to have an empty `{}` 

The previous implementation set the thresholds in the [set_limits()](https://github.com/Azure/WALinuxAgent/blob/master/azurelinuxagent/common/cgroups.py#L235) call, and the [cgroup is tracked](https://github.com/Azure/WALinuxAgent/blob/master/azurelinuxagent/common/cgroups.py#L236) (added to the common data structure _tracked). This was then referred by the monitor thread, but still, it did not work and was giving the `'NoneType' object has no attribute '__getitem__'` error - could be signaling that the cgroup in the [`_tracked`](https://github.com/Azure/WALinuxAgent/blob/master/azurelinuxagent/common/cgroups.py#L180) dictionary might not be populated in the first place with the right Cgroups objects (which has the thresholds assigned.

Reply:
The original line 429:
```
        self.threshold = None
```
Replace with:
```    
        cpu_limit = self.get_cpu_limits()
        mem_limit = self.get_memory_limits()
        self.threshold = {"cpu": cpu_limit, "memory": mem_limit}
```
These are default values, but you have no mechanism for defining non-default values anyway.

Another alternative: remove CGroups.threshold as a value and replace it with CGroups.threshold() as a method. When this function is called, the CGroups object computes thresholds right then and there. Your initial implementation would still return only defaults. A subsequent implementation (one which, for example, did file I/O to discover a per-extension setting) would only do that work when (and if) the threshold() method was called.

Reply:
Cleaned the implementation - Moved the limits into its [own class](https://github.com/Azure/WALinuxAgent/pull/1394/files#diff-0d947136d13fb1ee11231cfa47a36890R784), which can be extended in future if other new limits need to be added (IO, etc.) and enabled the plumbing of the limits to be set per CGroup.